### DISCUSSION TITLE
Laser tilt, polarization, ionization and related questions

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#511] [General]
Few questions for the experts i'd appreciate to find the answer on

1) In this example posted below at the end (taken from modified cone.deck) which runs on average PC just few minutes even on a single core, i have set the tilting of the laser but seems the way how to do that, which anyone can admit is really really tricky in EPOCH, was taken from 2D case. As a result the spot produced on the surface is not an ellipse but perfect circle. Any advise how to resolve that?

![_20230621_045733_Ei 0003    30 0fs 000](https://github.com/Warwick-Plasma/epoch/assets/112348247/1055c402-3c13-45da-8331-dfeaa3454fe3)

![_20230621_044057_Ei 0009    90 0fs 000](https://github.com/Warwick-Plasma/epoch/assets/112348247/94ae1811-80a6-4d1e-807b-edc5a8b08715)

2) In the same demo what was the polarization? Looks like a p-polarization, or is it s one? how to set different  polarization, including circular one?

3) **About ionization**  i still have many many many questions: In this demo case the initial Z was taken fixed. The initial number of particles is set by the npart_per_cell = 1 for each electron and ion (and the nparticles = nx * ny * nz * 0.3 is by default ignored).  Now suppose i would like to allow collisional ionization with initial  Z=1 how i have to change this demo deck for it to start working  ? 

Simply saying the only what needed with this demo deck is the charge started to be changing instead of being constant

/** (i will read it in diagnostics.f90 where it is defined as Z=current%charge with the current pointer defined as current => species_list(ispecies)%attached_list%head. Hence i do not need that diagnostics started to perform any outputs related to the charge species etc. Note that the OUTPUT block does not have "charge = always")

  **Other questions about ionization:**

  3a)  if to set, say, same npart_per_cell = 1 what will happen during simulations? Will the code try to keep npart_per_cell = 1  due to growing number of electrons by increasing weight OR  the number of particles per cell  will automatically grow potentially slowing the simulation 79 times in case of fully ionized gold? 
  3b) Does coll_n_step = 10 speeds up the simulation with the collisional ionization too ? Or only ci_n_step=10 doing so?
  3c) Sorry for my English but after reading this 100 times i still totally failed to understand what this means: "unique_electron_species - If “T”, this generates a unique electron species to populate with ejected electrons from each ion charge state. The user must use this, or ionisation_electron_species."
     What is "unique_electron_species" and why user needs to specify it or "ionisation_electron_species"  if no ionization information user needs  to be in the output? In simplest case what  needed is just to tell the code to start calculating ionization charge state Z instead of keeping it constant, something else is an extra, isn't it? If i use unique_electron_species=F then what will happen? If i use ionisation_electron_species=F then also what code will be doing? I totally do not understand what is ionisation_electron_species ? Was meant here ionisation of ion species as electrons can not be ionized?  And even if to allow this as a jargon to shorten this long keyword what was specifically meant here ? This confusion is because several working decks with ionization are clearly needed to understand what is what, so far i did not succeed to launch a single working example, all of them complain about something you are scratching the head while trying trying trying different variants:

**ERROR . Not all required elements of input deck specified.
 Please fix input deck and rerun code** 

  3d) The "identify" was mentioned as needed  but the deck snippet was missing electrons (in the paragraph which explains Ionization in the description  [https://epochpic.github.io/documentation/input_deck/input_deck_species.html](url) ) which just adds the difficulty for the user to assemble any working example by themselves. 

4) Every time i post the code here clicking Ctrl+e something wrong is always happening because GitGub distorts the text to look like a disaster. I now know how to fix that but always forget. Why GitHub does not  fix an obvious bug ? Or it is the "feature"   :)? 

The demo input.deck :
 
```begin:control
  nx =  80
  ny =  160
  nz =  160

 nparticles = nx * ny * nz * 0.3

# Final time of simulation
  t_end = 50 * femto

# Size of domain
  x_min = -6 * micron
  x_max =  2 * micron
  y_min = -8 * micron
  y_max =  8 * micron
  z_min = -8 * micron
  z_max =  8 * micron

end:control

#-------------------------------------
begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = simple_outflow
  bc_y_max = simple_outflow
  bc_z_min = simple_outflow
  bc_z_max = simple_outflow
end:boundaries

#-----------------------------------------------------------
begin:constant
  laser_wavelength = 0.8 * micron
  omega = 2 * pi * c / laser_wavelength

  n_crit = critical(omega)
  # density (Fused Silica at full ionisation: 6.6e29)
  ne_max = 30 * n_crit

  laser_k = 2 * pi / laser_wavelength
  den_solid = 0.6875 * n_crit
  SiCharge = 12
  Ocharge  = 6

  ladder_step = -0.1 * micron

  r = sqrt(y^2 + z^2)

  laser_angle = pi / 4
  laser_fwhm = 5 * micron
  laser_dur_fwhm = 50.e-15

end:constant

 begin:collisions
  use_collisions = T
  use_nanbu = T
  coulomb_log = auto
#  collisional_ionisation = T
#  collide = electron electron on     #this key to switch off specific collisions because USE COLLISIONS=T means they are all ON by default (besides coll.ionization which needs additional switch to turn ON)
  coll_n_step = 10
  collide = all
end:collisions

#------------------------------------------------
begin:species
  name = Silicon
  charge = SiCharge  
  mass   = 1836.2*28.09

number_density = if( (x lt 0.3 * micron)    and (x gt 1. *ladder_step), den_solid, 0.)
number_density = if( (x lt 1. *ladder_step) and (x gt 2. *ladder_step), den_solid*0.9,  number_density(Silicon))
number_density = if( (x lt 2. *ladder_step) and (x gt 3. *ladder_step), den_solid*0.8,  number_density(Silicon))
number_density = if( (x lt 3. *ladder_step) and (x gt 4. *ladder_step), den_solid*0.7,  number_density(Silicon))
number_density = if( (x lt 4. *ladder_step) and (x gt 5. *ladder_step), den_solid*0.6,  number_density(Silicon))
number_density = if( (x lt 5. *ladder_step) and (x gt 6. *ladder_step), den_solid*0.5,  number_density(Silicon))
number_density = if( (x lt 6. *ladder_step) and (x gt 7. *ladder_step), den_solid*0.4,  number_density(Silicon))
number_density = if( (x lt 7. *ladder_step) and (x gt 8. *ladder_step), den_solid*0.3,  number_density(Silicon))
number_density = if( (x lt 8. *ladder_step) and (x gt 9. *ladder_step), den_solid*0.2,  number_density(Silicon))
number_density = if( (x lt 9. *ladder_step) and (x gt 10.*ladder_step), den_solid*0.1,  number_density(Silicon))
number_density = if( (x lt 10.*ladder_step) and (x gt 12.*ladder_step), den_solid*0.05, number_density(Silicon))


#  number_density = if( (x gt 0.) and (x lt target_x_outer), den_solid, 0.)
#  number_density = if( (x lt 0.) and (x gt target_x_wapor_inner), den_solid *(1+x/target_x_wapor_inner), number_density(Silicon))

   number_density_min = 0.05 * n_crit/SiCharge
#  density_max = ne_max
   npart_per_cell = 1
end:species
#---------------------------------------------------
begin:species
  name = Oxygen
  charge = Ocharge
    mass = 1836.2*16.09

  number_density = number_density(Silicon) * 2

  number_density_min = 0.05 * n_crit/Ocharge
# density_max = ne_max
  npart_per_cell = 1
end:species

#---------------------------------------------------
begin:species
  name = electron
  charge = -1.0
  mass = 1.0
  number_density = number_density(Silicon)*SiCharge  + number_density(Oxygen)*Ocharge

  npart_per_cell = 1
  number_density_min = 0.01 * n_crit

end:species

#--------------------------------------------------

begin:laser
  boundary = x_min

  intensity_w_cm2 = 2.0e18

  profile = gauss(r, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))
  t_profile = gauss(time, laser_dur_fwhm, laser_dur_fwhm / (2.0 * sqrt(loge(2.0))))

  phase = -y * laser_k * sin(laser_angle)          # Vary wave-front position in space
  lambda = laser_wavelength * cos(laser_angle)     # Space wave-fronts apart  # correctly along k

end:laser

#====================== O U T P U T =========================

begin:output_global
  force_final_to_be_restartable = T
end:output_global


begin:output
  name = normal
  file_prefix = x

# Simulated time between output dumps
  dt_snapshot = 10 * femto

# Properties on grid
  grid = always
  ey = always
  average_particle_energy = always
  number_density = always + species
  temperature = always + species

  # Extended io
  distribution_functions = always
end:output ```

**Comment by Status-Mirror** [comment]
Hey Dan, I'll do my best to answer your questions as simply as I can.

1. Your profile is set to `profile = gauss(r, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))`, which has axial symmetry about `r=0`. If you want to focus to an ellipse, you could set your profile to something like `profile = gauss(y,0,y_size) * gauss(z,0,z_size)`, which would let you have different focal spot sizes in the `y` and `z` directions.
2. On polarisation, check the `pol_angle` parameter in the [laser block](https://epochpic.github.io/documentation/input_deck/input_deck_laser.html). The default polarisation on the `x_min` boundary is along `Ey`. Because your tilt is in the y-direction (`phase = -y * laser_k * sin(laser_angle)`), then this is P-polarisation. For S-polarisaiton, add the line `pol_angle=0.5*pi` to the laser block. **Edit**: For circular polarisation, you could use two laser blocks which overlap in space, but have perpendicular polarisations and a phase difference (this may not work with the tilted-field trick, and it may be better to rotate your target and inject a laser normal to the boundary).
3. On ionisation: 
**a)** `npart_per_cell` is only used for setting up the simulation, the code does **not** keep this number fixed. When ionisation occurs, more macro-electrons are added to the simulation window. The more ionisation which occurs, the slower the code will run.
**b)** Despite sharing an input deck block, EPOCH treats collisions (elastic scatter between charged particles) and collisional ionisation (incident electrons ionising atoms and ions) as separate processes. Calculations are performed independently - hence, `coll_n_steps` speeds up collisions, and `ci_n_steps` speeds up collisional ionisation.
**c)** In 3a), I say that ionisation creates new macro-electrons. But we need to tell EPOCH which species to put these new macro-electrons into. If you use `unique_electron_species`, then EPOCH creates a separate species for each ionisation state. For example, when we ionise from `Oxygen3` to `Oxygen4` the released macro-electron would be added to the `electron_from_Oxygen3` species, which is automatically created when `unique_electron_species=T`. If you set  `ionisation_electron_species = electron`, then all new macro-electrons are added to the `electron` species.
**d)** I agree it would be wise to mention the `identify` line in the code snippet for the ionisation section. I'll add this to the documentation.
4. I suppose GitHub isn't smart enough to realise when something is supposed to be a code or not. Maybe AI will rescue us in the future :P

Hope this helps,
Stuart

**Comment by DanRRRR** [comment]
Thanks Stuart for the detailed answers. 

The solution for the ellipse surprised me though. It is very good case for potential confusion 

I thought the crossection of beam is circular (if cut it perpendicular to the beam axis) and the code will exhibit  ellipsicity somehow automatically based on the whatever final angle on the target will be. For example we may start the laser at 45 degrees but when it propagates in the matter it may wildly refract and land on the target at some very different angle we do not know in advance. The ellipse parameter of course  will be different in this case and we do not know how to setup it. 

But looks like in some initial points the beam has to be set as an ellipse because crossection has to be taken perpendicular not to the beam axis but perpendicular to the x axis

As to ionization i need to find why my small add-on to diagnostics subroutine now crashes the code with ionization, while it did not crash without ionization

**Reply by Status-Mirror** [reply]
Hey @DanRRRR 

I think I misunderstood your request regarding ellipticity - I thought you wanted to inject your laser with an initial elliptical profile.

EPOCH boundaries are designed assuming lasers are injected perpendicular to the boundary - on `x_min`, the laser routines can only set up lasers where $\pmb{E} = E_y \pmb{\hat{y}} + E_z \pmb{\hat{z}}$. We can only create "tilted" lasers in a few cases (e.g. plane waves) by manipulating phase fronts to trick the EPOCH field solver into propagating angled beams. There are cases where this trick doesn't work - I have never seen anyone successfully inject an angled Gaussian beam.

In all cases with a single laser, it is easier to rotate your target and inject the beam normally.

I can't really help users with custom EPOCH code, but I can have a guess. Your field diagnostics shouldn't be changing with ionisation, as these routines don't add or modify any EM fields. Ionisation will, however, introduce new particles. Maybe your diagnostics are sensitive to the number of macro-particles in the simulation?

**Comment by DanRRRR** [comment]
Stuart, I know this is not easy  :)  In some  cases also focusing of the beam is used with specific D/F and Rayleigh lengths. Some other codes i tried failed at that why I used EPOCH because saw in description  this problem was solved here. 

In this demo clearly on the top of the simulation box initial distribution of the field will resemble ellipse because crossection of the beam is purely circular and we  use 45 degrees angle. Will your solution do the trick at least as a simplest approximation ? Of course in more complex cases where laser besides being tilted also is focusing on the target pretty tightly this might now work.

Tilting the target would be probably a solution but also could be a headache in some cases...

One more short question: is this demo case above restartable from any SDF or only from the last one? This part is the most confusing in the description. Clearly these regular SDFs in my demo have not enough info for restart and only potentially "force_final_to_be_restartable = T" creates one. BTW, the EPOCH description does not say how specifically to start it as a continuation and if some additional files in the dir created besides SDF (like normal.visit etc) are needed for that. Are they really needed for restart? It is almost impossible to be sure that for continuation you included into large SDF everything.  But seems the "force_final_to_be_restartable = T" which i suppose automatically includes everything needed can be done only ones at the end of simulation making it useless in case of crash. Did i correctly understand why i so far got no answer on Continuation post #480 and got only polite sending into the space to RTFM :). My impression is that due to that complexity most people simply do not use continuation and rely on the run till the victory end 

In summary, here are the questions about restart
- Please provide at least a single example of restart from intermedite time (suppose we had outage). Take  any examples of input.decks, for example code.deck
- Are cone.deck from example_deck restartable from **large*** outputs ? Are large outputs designed specifically for restart? Then what minimum requirement of output has to be included there?
- Are input.deck, restart.visit, normal necessary to restart from any intermediate time ?
- Can we change something in the middle of simulation and restart from intermediate output ?
- Can number of cores be changed when continue run from restart ?

**Comment by DanRRRR** [comment]
Stuart, Thanks for adding large paragraph about restart. Great job

The linear density gradient above was done manually by multiple fixed steps  because any automatic ways via  loops i tried failed. The defects were that the layers were not fully filled with the particles etc.  If there exist any better elegant way to set the gradients i would also appreciate to learn that.
